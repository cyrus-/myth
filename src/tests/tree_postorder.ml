let s = "
type nat =
  | O
  | S of nat

type list =
  | Nil
  | Cons of nat * list

type tree =
  | Leaf
  | Node of tree * nat * tree

let rec append (l1:list) (l2:list) : list =
  match l1 with
  | Nil -> l2
  | Cons (x, l1) -> Cons (x, append l1 l2)
;;

let tree_postorder : tree -> list |>
{  Leaf => []
| Node (Leaf, 1, Leaf) => [1]
| Node (Leaf, 2, Leaf) => [2]
| Node (Node (Leaf, 1, Leaf), 2, Leaf) => [1;2]
| Node (Leaf, 1, Node (Leaf, 2, Leaf)) => [2;1]
| Node (Node (Leaf, 1, Leaf), 0, Node (Leaf, 2, Leaf) ) => [1;2;0]
| Node (Node (Leaf, 2, Leaf), 0, Node (Leaf, 1, Leaf) ) => [2;1;0]
| Node (Node (Node (Leaf, 2, Leaf), 0, Node (Leaf, 1, Leaf) ), 0, Leaf) => [2;1;0;0]
| Node (Leaf, 2, Node (Node (Leaf, 2, Leaf), 0, Node (Leaf, 1, Leaf) )) => [2;1;0;2]
} = ?"
