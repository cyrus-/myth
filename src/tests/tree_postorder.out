(* Generated Thu Nov 13 00:37:25 2014 *)
let tree_postorder : tree -> list =
  let rec f1 (t1:tree) : list =
    match t1 with
      | Leaf -> Nil
      | Node (t2, n1, t3) -> (match f1 t2 with
                                | Nil -> append (f1 t3) (Cons (n1, Nil))
                                | Cons (n2, l1) -> Cons (n2,
                                                     append
                                                       (append l1 (f1 t3))
                                                       (Cons (n1, Nil))))
  in
    f1
;;
